doc/dev/libraries.md

```diff
@@ -0,0 +1,84 @@
+  Specifically targets C and not intended for use via bindings.
+
+  **Prefix:** `elektra*`
+
+- `libelektra-highlevel-c`:
+  The C high-level API for reading/writing configuration with Elektra.
+  Intended for use by applications.
+  Partially intended for use with code-generation.
+  Specifically targets C and not intended for use via bindings.
+
+  **Prefix:** `elektraHlc*` (better suggestions welcome)
+  (Note: Even though there isn't any other `highlevel` implementation, I added the `-c` suffix to show this lib is only for C)
+
+- `libelektra-opts`:
+  The API for parsing command-line arguments according to Elektra's spec.
+  **Prefix:** `elektraOpts*`
+- `libelektra-notification`:
+  The API for setting up callbacks and automatically updated variables linked to changes in the KDB.
+
+  **Prefix:** `elektraNotification*`
+
+- `libelektra-io`:
+  Elektra's API for asynchronous operations.
+
+  **Prefix:** `elektraIo*`
+
+- `libelektra-merge`:
+  The API for merging and/or detecting conflicts between two keysets.
+
+  **Prefix:** `elektraMerge`
+
+- `libelektra-pluginbase`:
+  The API _used by_ plugins.
+
+  **Prefix:** `elektraPlugin*`
+  (Note: previously `libelektra-plugin`)
+
+- `libelektra-plugin`:
+  The API for _interacting with_ plugins.
+  This includes loading plugins and calling exported functions.
+
+  **Prefix:**: `elektraPlugin*`
+  (Note: previously `libelektra-invoke` & `loader`)
+
+- `libelektra-mount`:
+  The API for manipulating mountpoints.
+
+  **Prefix:** `elektraMount*`
+
+- `libelektra-type`:
+  The API that defines Elektra's type system. (Alternative: The API that defines the type system used by `libelektra-highlevel-c`.)
+
+  **Names:** `elektra<TYPE>ToString` and `elektraKeyTo<TYPE>`
+  (Note: extracted from `libelektra-ease`)
+
+- `libelektra-ease`:
+  A collection of various other APIs that help when interacting with `ElektraKey` and `ElektraKeyset`.
+  Contains APIs that are not large, specific, or complex enough to justify a separate library and not minimal enough for `libelektra-core`, but can be implemented on top of the public API.
+  The APIs should not specifically target C and should be usable via bindings (if appropriate for the other language).
+
+  **Prefix:** `elektra*`
+  (Note: rest of old `libelektra-ease`, merged with `libelektra-meta` & `libelektra-utility`; cleanup needed)
+
+- `libelektra-operations`:
+  Like `libelektra-ease`, this library contains various additional `ElektraKey` and `ElektraKeyset` APIs beyond the minimal API.
+  However, these APIs specifically need access to private APIs of `libelektra-core` to be implemented (efficiently).
+  The APIs should not specifically target C and should be usable via bindings (if appropriate for the other language).
+  All APIs must abstract over unstable details of private APIs and provide a stable API, or **very clearly** state the stability restrictions.
```

Comment:
> if a function has restrictions on stability, these must be clearly stated. This could simply be

It cannot be on function level. Maintainers expect that libraries with the same SOVERSION **are** compatible. So we can only have libraries that might be incompatible (SOVERSION might get pumped from time to time).

For things that are refactored by you, I wouldn't keep the option to pump SOVERSION as this is a **major** annoyance to users. Users want stability. They want their applications to continue working. This is our #1 goal, see doc/GOALS.md

> That's why https://github.com/ElektraInitiative/libelektra/pull/4246 proposes different headers. Then users need to explicitly opt in to non-stable APIs.

I agree that it must be very clear for users if there are any non-stable APIs. Can you clarify this in the decisions, please? Ideally we don't have any unstable libraries, though. Thus we really should watch out to make everything as simple&minimal as possible. Otherwise it would be impossible to make a beautiful API which lasts long.

Reply:
It's not as simple as "this library is stable/unstable". We already agreed that e.g. `struct _Key` cannot be fully private (needed by `mmap`, will be used directly by `libelektra-operations` etc.), but also can't really be stable API. If we made it stable API there would be no point to hide the layout at all and we could just let users access `struct _Key` directly.

So their are certain things that are technically public (i.e. exported), but we don't consider public API and therefore don't provide stability guarantees.

A similar example would be

```c
// in elektra/foo.h
elektraFooMainFunc();

// in elektra/foo/unstable.h
elektraFooDoSomething();
elektraFooDoMoreStuff();
```

the public API of `libelektra-foo` is just `elektraFooMainFunc()`. If you only call that you get all the stability guarantees you expect.

However, for whatever reason (e.g. special use by other libs, or maybe new APIs that need more testing before we make them stable), `libelektra-foo` also exports the functions `elektraFooDoSomething()` and `elektraFooDoMoreStuff()`. Probably, these are part of the implementation of `elektraFooMainFunc()`. If you call `elektraFooDoSomething()` or `elektraFooDoMoreStuff()`, things may change, things may break, the functions may even be removed entirely, you need to deal with that.

Of course, if there is public API in another library of Elektra e.g.

```c
// in elektra/bar.h
elektraBarMainFunc();
```

that calls `elektraFooDoSomething()` or `elektraFooDoMoreStuff()` then you can call `elektraBarMainFunc` without any issues. `elektraBarMainFunc` is public stable API and we will make sure things don't break, no matter how it is implemented and what other APIs it calls.

> Users want stability. They want their applications to continue working. 

And that's why 99.999% of applications should only ever use the stable public API. But if you can tie your application to a specific version of Elektra (e.g. fully integrated embedded system) or your application isn't public and will always be updated when things break, then you could use the unstable APIs too.

> we really should watch out to make everything as simple&minimal as possible

"as possible" is doing a lot of work in that sentence... What is possible?

IMO the core should be small, so that when somebody needs it they can go to the bare minimum.

But for most other things, I think usability and developer experience is much more important than being minimal. No one will use a library, where you have to make 6 API calls to do a common task, when there could just be a single API. If the library get's used at all, people will create wrapper libraries to simplify things and that is IMO a failure of API design in the underlying library.

> beautiful API

APIs don't need to be beautiful; they need to be functional, usable, safe, etc. But I get want you mean.

---

With this system problems only come up in two cases:
1. You mix and match Elektra's `.so` files between `x.y.z` versions. That's clearly a broken install and shouldn't be supported.
2. Some external library/application depends on Elektra and used an unstable API, but wasn't updated for the newest release. That's a problem with the library/application and the author should avoid the unstable APIs.

Reply:
> It's not as simple as "this library is stable/unstable".

Then **we** need to make it simpler. In the end of the day maintainers/application developer trust that libraries without changes in SONAME are compatible. Any other rules we might have can only be of internal nature (within the Elektra initiative).

> We already agreed that e.g. struct _Key cannot be fully private (needed by mmap, will be used directly by libelektra-operations etc.)

- in case of mmap cache files there is version information. Caches simply get discarded when `struct _Key` changes.
- libelektra-operations (and probably also other libs+plugins like "cache") is simply bound to be exactly the same version as libelektra-core. This can easily be done in package managers as they are both from the same source package.

> "as possible" is doing a lot of work in that sentence... What is possible?

Remove/Cleanup as much as possible for 1.0, see  #2987

> create wrapper libraries to simplify things and that is IMO a failure of API design in the underlying library.

I don't think it is a big deal in general. There are often some requirements like how to do threading, different naming conventions etc. that require wrappers. I often wrap KDB, parentKey and main KeySet to one struct to pass it around more easily. The wrapping code is super-small and written in a few minutes, so I would not worry about that for our low-level APIs. Only in more high-level APIs people of course expect that simple things (like getting one value) can be done with a few calls.

> With this system problems only come up in two cases:

There is a third:

3. We have stable libraries that exactly follow semantic versioning/SONAME (no breaks **at all**, i.e. ONLY what symbol versioning allows). We have other libs that are a bit more experimental. If people use them, they might get problems. We don't encourage to include such header files or to link against such libraries until we get them out of the experimental stage.

Reply:
> There are often some requirements like how to do threading, different naming conventions etc. that require wrappers.

Those are fine. These are things that Elektra leaves up to the user intentionally. The bad wrappers are those that work around shortcomings, inconveniences and other issues in Elektra's API.

> I often wrap KDB, parentKey and main KeySet to one struct to pass it around more easily.

If you do something like that _often_, then I would consider that a serious problem with Elektra's API. But in this case we already have a solution. The highlevel API has exactly this struct.

> The wrapping code is super-small and written in a few minutes

But every time somebody write one of these wrappers they make additional assumptions and muddy the use case of Elektra. It also makes Hyrum's Law worse.

> Then **we** need to make it simpler.

That's not how things work. We want to keep `struct _Key` private, but we want to have stuff outside `libelektra-core` that as access to `struct _Key`. Ergo we need to make `struct _Key` technically public, but cannot guarantee that it remains the same.

> In the end of the day maintainers/application developer trust that libraries without changes in SONAME are compatible.

Sure, but that's easy to fix. We just need to make sure that our SONAME changes when we break something. The easiest (but least useful) way is to include the full x.y.z version in the SONAME. 

> There is a third: [...]

Not sure how what you described is a problem that could happen with my proposed header/versioning system. IMO what you describe is basically the same as what I proposed by less detailed.

> libelektra-operations (and probably also other libs+plugins like "cache") is simply bound to be exactly the same version as libelektra-core.

I'd say this should apply to everything.

So the rules would be:

1. _Everything_ in this repo will be released together with the same version numbers.
2. Semantic versioning guarantees only apply between this repo and the outside, but not within this repo. An application built for `libelektra-core` v1.0.0 will work with `libelektra-core` v1.x.y. But any `libelektra-ABC` va.b.c _only works_ with `libelektra-XYZ` va.b.c. In other words, all of Elektra's packages must be installed with the same version. Like you said, any proper package manager setup would already enforce this.
3. Some libraries have headers `elektra/foo/unstable.h` that are exempt from semantic versioning.

Reply:
> We want to keep struct _Key private, but we want to have stuff outside libelektra-core that as access to struct _Key.

It might be outside of core but inside of Elektra.

> Ergo we need to make struct _Key technically public, but cannot guarantee that it remains the same.

No, we do not "need to". We could and probably should not even install the header describing how _Key looks like.

> Sure, but that's easy to fix. We just need to make sure that our SONAME changes when we break something. The easiest (but least useful) way is to include the full x.y.z version in the SONAME.

Exactly.

>  I'd say this should apply to everything.

Yes, I agree. An inside/outside Elektra world-view simplifies everything a lot.

> Everything in this repo will be released together with the same version numbers.

:+1:  Can you please add this do doc/VERSION.md?

> Some libraries have headers elektra/foo/unstable.h that are exempt from semantic versioning.

See also  #4245. We might have unstable libraries with SOVERSION pumps from time to time (ideally we won't but this is a compromise I am willing to take) but we would only create (huge) troubles if we do stuff nobody expects. These "unstable.h" things also do not solve a real problem. As long as we do not know how to offer some functionality, it is better not to offer it, and especially not in our core libraries.

Reply:
> See also https://github.com/ElektraInitiative/libelektra/pull/4245.

This is #4245... I assume you mean [this comment](https://github.com/ElektraInitiative/libelektra/pull/4245#issuecomment-1247818411).

As stated in [my response](https://github.com/ElektraInitiative/libelektra/pull/4245#issuecomment-1247923112), let's focus on the prefix and library split question first.